src/main.rs

```diff
@@ -166,52 +176,119 @@ mod test_song_segments {
 		let path = path.as_ref();
-		let file_name = path.file_name().unwrap().to_str().unwrap().to_string();
-		let name_split = file_name.split('_').collect::<Vec<_>>();
-		let song_id = name_split[1].to_string();
-		let song_segment_split = name_split[2].split(".").collect::<Vec<_>>();
-		let song_segment_id = song_segment_split[0].to_string();
-		let song_segment_format = song_segment_split[1].to_string();
-		let song = songs.entry(song_id.clone()).or_insert(Song {
-			id: song_id,
-			segments: HashMap::<String, SongSegment>::new(),
-			has_end: false,
-			has_multiple_loops: false,
-			has_dedicated_transitions: false
-		});
-		if !song.has_end && song_segment_id == "end"{
-			song.has_end = true;
-		}
-		if !song.has_multiple_loops && song_segment_id != "loop" && REGEX_IS_LOOP.is_match(&song_segment_id) {
-			song.has_multiple_loops = true;
-		}
-		if !song.has_dedicated_transitions && REGEX_IS_DEDICATED_TRANSITION.is_match(&song_segment_id) {
-			song.has_dedicated_transitions = true;
-		}
-		if song.segments.contains_key(&song_segment_id) {
-			// Panic here, because having multiple files with the same ID is ambiguous
-			panic!(format!("Found multiple segments with same ID: Song: {} Segment: {}", song.id, song_segment_id))
+		match detect_file_type(path){
+			FileFormat::SegmentFormat => {
+				let song_id = get_song_name(path.file_name().unwrap().to_str().unwrap());
+				let segment = parse_segment(path.file_name().unwrap().to_str().unwrap()).unwrap();
+				let song = songs.entry(song_id.to_string()).or_insert(Song {
+					id: song_id,
+					segments: HashMap::<String, SongSegment>::new(),
+					has_end: false,
+					has_multiple_loops: false,
+					has_dedicated_transitions: false,
+					is_archive: false,
+				});
+				if segment.id == "end"{
+					song.has_end = true;
+				}
+				if segment.id != "loop" && REGEX_IS_LOOP.is_match(&segment.id) {
+					song.has_multiple_loops = true;
+				}
+				if REGEX_IS_DEDICATED_TRANSITION.is_match(&segment.id) {
+					song.has_dedicated_transitions = true;
+				}
+				if song.segments.contains_key(&segment.id.to_string()) {
+					// Panic here, because having multiple files with the same ID is ambiguous
+					panic!(format!("Found multiple segments with same ID: Song: {} Segment: {}", song.id, segment.id))
+				}
+				song.segments.entry(segment.id.to_string()).or_insert(segment);
+			}
+		    FileFormat::SongArchiveFormat => {
+				let archive = ZipArchive::new(File::open(path).unwrap()).unwrap();
+				let song_id = get_song_name(path.file_name().unwrap().to_str().unwrap());
+				println!("Encountered Archive {}", song_id);
+				let song = songs.entry(song_id.to_string()).or_insert(Song {
+					id: song_id,
+					segments: HashMap::<String, SongSegment>::new(),
+					has_end: false,
+					has_multiple_loops: false,
+					has_dedicated_transitions: false,
+					is_archive: true,
+				});
+				for segment_path in archive.file_names(){
+					let segment_name = format!("song_archive_{}",segment_path);
```

Comment:
If there are 2 songs in archive format, they will conflict.
```suggestion
					let segment_name = format!("song_{}_{}", song.id, segment_path);
```

Reply:
No reason to not commit this change, but notable that the addition of `song_archive_` is only so that it is properly accepted by `parse_segment()`, which expects but does not use  the `archive` part of `song_archive` is thrown away afterwards.

It actually might be better to reformat parse_segment to use `name_split.last()` instead of parsing past two unneeded fields...

Reply:
Agree, as discussed in #8

Reply:
most recent commit should be a more elegant solution, and should also reduce possible parsing errors.